### PR TITLE
type - initial API for type normalization

### PR DESCRIPTION
**Pull request by pounard**

### CODE REVIEW

File: BENCHMARK.md [hunk 1]
```diff
@@ -24,6 +24,36 @@ Run benchmark and compare with tag:
...
+| ConversionBench | benchRamseyUuidToSqlNullType |     | 3000 | 5   | 1.987mb  | 79.733μs  | ±0.54% |
+| ConversionBench | benchRamseyUuidToSqlWithType |     | 3000 | 5   | 1.988mb  | 42.296μs  | ±0.82% |
+| ConversionBench | benchArrayFromSql            |     | 3000 | 5   | 2.016mb  | 304.989μs | ±0.69% |
+| ConversionBench | benchArrayToSql              |     | 3000 | 5   | 2.016mb  | 230.059μs | ±0.80% |
+| ConversionBench | benchArrayToSqlNullType      |     | 3000 | 5   | 2.016mb  | 240.368μs | ±0.93% |
+| ConversionBench | benchArrayToSqlWithType      |     | 3000 | 5   | 2.015mb  | 201.752μs | ±1.01% |
++-----------------+------------------------------+-----+------+-----+----------+-----------+--------+
+```
+
+Analysis: added the type API, induces some mandatory slowness in value conversion. Some deviations
```
Comment (Lonnytunes) [inline BENCHMARK.md:53]:
```suggestion
Analysis: adding the type API induces some mandatory slowness in value conversion. Some deviations
```

File: src/Converter/Converter.php [hunk 2]
```diff
@@ -244,9 +250,13 @@ protected function getConverterContext(): ConverterContext
     /**
      * Run all plugins to convert a value.
      */
-    protected function fromSqlUsingPlugins(null|int|float|string|object $value, string $type, ?string $realType = null): mixed
+    protected function fromSqlUsingPlugins(null|int|float|string|object $value, ?string $type, ?string $realType = null): mixed
     {
         $realType ??= $type;
+        if (!$realType) {
+            throw new QueryBuilderError();
```
Comment (Lonnytunes) [inline src/Converter/Converter.php:257]: No message?

File: src/Converter/Converter.php [hunk 3]
```diff
@@ -278,9 +288,9 @@ protected function fromSqlDefault(string $type, null|int|float|string|object $va
     /**
      * Run all plugins to convert a value.
      */
-    protected function toSqlUsingPlugins(mixed $value, string $type, ?string $realType = null): null|int|float|string|object
+    protected function toSqlUsingPlugins(mixed $value, ?Type $type, ?Type $realType = null): null|int|float|string|object
     {
-        $realType ??= $type;
+        $realType ??= $type ?? throw new QueryBuilderError();
```
Comment (Lonnytunes) [inline src/Converter/Converter.php:293]: No error message here as well?

File: src/Type/InternalType.php [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    case DECIMAL;
+
+    /** Floating point numner. */
+    case FLOAT;
+
+    /** Big floating point numner. */
+    case FLOAT_BIG;
+
+    /** Small floating point numner. */
+    case FLOAT_SMALL;
```
Comment (Lonnytunes) [inline src/Type/InternalType.php:42]:
```suggestion
    /** Floating point number. */
    case FLOAT;

    /** Big floating point number. */
    case FLOAT_BIG;

    /** Small floating point number. */
    case FLOAT_SMALL;
```

File: src/Type/InternalType.php [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+    public function isNumeric(): bool
+    {
+        return
+            $this === self::DECIMAL ||
+            $this === self::FLOAT || $this === self::FLOAT_BIG || $this === self::FLOAT_SMALL ||
+            $this === self::INT || $this === self::INT_BIG || $this === self::INT_SMALL ||
+            $this === self::IDENTITY || $this === self::IDENTITY_BIG || $this === self::IDENTITY_SMALL ||
+            $this === self::SERIAL || $this === self::SERIAL_BIG || $this === self::SERIAL_SMALL
+        ;
```
Comment (Lonnytunes) [inline src/Type/InternalType.php:279]: Would you not prefer to use `\in_array()` ?

File: src/Type/Type.php [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+
+    public function toArray(): self
+    {
+        $ret = clone $this;
+        $ret->array = true;
+
+        return $ret;
+    }
+
+    public function toNoArray(): self
```
Comment (Lonnytunes) [inline src/Type/Type.php:425]:
```suggestion
    public function toNonArray(): self
```

File: src/Converter/Converter.php (hunk 2 above)
Comment (pounard) [reply src/Converter/Converter.php:257]: Really there's no need here, this is internal code that isn't supposed to be called from outside: the error case would simply mean the calling code is wrong.

File: src/Converter/Converter.php (hunk 3 above)
Comment (pounard) [reply src/Converter/Converter.php:293]: Same as upper.

File: src/Type/InternalType.php (hunk 5 above)
Comment (pounard) [reply src/Type/InternalType.php:279]: Really not, I'm afraid `\in_array()` would be slower, although now that I have benchmarks, I could test it!

File: src/Type/Type.php (hunk 6 above)
Comment (pounard) [reply src/Type/Type.php:425]: Good idea, I was really unhappy with this name, but I can't simply apply your suggestion I need to fix core everywhere it's called.

Comment (pounard) [reply src/Type/Type.php:425]: Done !

**Pull request merged**